File: solutions/c/raindrops/2/raindrops.c
#include "raindrops.h"
#include <string.h>
#include <stdio.h>

typedef struct {
    int divisor;
    char* sound;
} Raindrop;

void convert(char result[], int number) {
    Raindrop drops[] = {
        {3, "Pling"},
        {5, "Plang"},
        {7, "Plong"}
    };

    result[0] = '\0';

    for (int i = 0; i < 3; i++) {
        if (number % drops[i].divisor == 0) {
            strcat(result, drops[i].sound);
        }
    }

    if (result[0] == '\0') sprintf(result, "%d", number);
}

File: solutions/haskell/raindrops/2/src/Raindrops.hs
module Raindrops (convert) where

sounds :: [(Int, String)]
sounds = [(3, "Pling"), (5, "Plang"), (7, "Plong")]

convert :: Int -> String
convert n
  | null result = show n
  | otherwise   = result
  where
    result = concatMap (check n) sounds

check :: Int -> (Int, String) -> String
check n (divisor, sound)
  | n `mod` divisor == 0 = sound
  | otherwise            = ""
